Improve RPATH handing in CoreNeuron/ReportingLib

Latest [build](https://bbpcode.epfl.ch/ci/job/hpc.neuron.toolchain/platform=bg/358/console) has failure message : 

```
-- Installing: /gpfs/bbp.cscs.ch/scratch/gss/bgq/kumbhar/SPACK_INSTALLS/SPACK_BBP_PREFIX/bg/install/2017_08_27_22_12/bgq-cnk-ppc64/xl-12.1/coreneuron-perfmodels-mlhhm5l6a4xbeldue4xktrpfem55r7um/lib64/libcoreneuron.a
-- Installing: /gpfs/bbp.cscs.ch/scratch/gss/bgq/kumbhar/SPACK_INSTALLS/SPACK_BBP_PREFIX/bg/install/2017_08_27_22_12/bgq-cnk-ppc64/xl-12.1/coreneuron-perfmodels-mlhhm5l6a4xbeldue4xktrpfem55r7um/bin/coreneuron_exec
CMake Error at apps/cmake_install.cmake:53 (file):
  file RPATH_CHANGE could not write new RPATH:

    /gpfs/bbp.cscs.ch/scratch/gss/bgq/kumbhar/SPACK_INSTALLS/SPACK_BBP_PREFIX/bg/install/2017_08_27_22_12/bgq-cnk-ppc64/xl-12.1/coreneuron-perfmodels-mlhhm5l6a4xbeldue4xktrpfem55r7um/lib:/gpfs/bbp.cscs.ch/scratch/gss/bgq/kumbhar/SPACK_INSTALLS/SPACK_BBP_PREFIX/bg/install/2017_08_27_22_12/bgq-cnk-ppc64/xl-12.1/coreneuron-perfmodels-mlhhm5l6a4xbeldue4xktrpfem55r7um/lib64:/bgsys/drivers/ppcfloor/comm/xl/lib:/gpfs/bbp.cscs.ch/scratch/gss/bgq/kumbhar/SPACK_INSTALLS/SPACK_BBP_PREFIX/bg/install/2017_08_27_22_12/bgq-cnk-ppc64/xl-12.1/neurodamus-develop-ysvskstb7enjybby3lkg6wccb2f75n3q/lib:/gpfs/bbp.cscs.ch/scratch/gss/bgq/kumbhar/SPACK_INSTALLS/SPACK_BBP_PREFIX/bg/install/2017_08_27_22_12/bgq-cnk-ppc64/xl-12.1/reportinglib-develop-egg74axinani2dbxeth6zm2w54h4uvjs/lib64

  to the file:

    /gpfs/bbp.cscs.ch/scratch/gss/bgq/kumbhar/SPACK_INSTALLS/SPACK_BBP_PREFIX/bg/install/2017_08_27_22_12/bgq-cnk-ppc64/xl-12.1/coreneuron-perfmodels-mlhhm5l6a4xbeldue4xktrpfem55r7um/bin/coreneuron_exec

  No valid ELF RPATH or RUNPATH entry exists in the file;
Call Stack (most recent call first):
  cmake_install.cmake:38 (include)


make: *** [install] Error 1
==> Executing phase : 'install'
++ spack find coreneuron@perfmodels +mpi %xl '^mpich'
```

ReportingLib and CoreNEURON cmake needs to be revisited.

--- a/packages/coreneuron/package.py
+++ b/packages/coreneuron/package.py
@@ -27,7 +27,7 @@ class Coreneuron(CMakePackage):
     url      = "https://github.com/BlueBrain/CoreNeuron"
     bbpurl   = "ssh://bbpcode.epfl.ch/sim/coreneuron"
 
-    version('develop', git=url, branch='fix_bgq', preferred=True)
+    version('develop', git=url, preferred=True)
 
     # TODO: same as develop but for legacy reasons
     version('perfmodels', git=url)

--- a/packages/coreneuron/package.py
+++ b/packages/coreneuron/package.py
@@ -27,7 +27,7 @@ class Coreneuron(CMakePackage):
     url      = "https://github.com/BlueBrain/CoreNeuron"
     bbpurl   = "ssh://bbpcode.epfl.ch/sim/coreneuron"
 
-    version('develop', git=url, preferred=True)
+    version('develop', git=url, branch='fix_bgq', preferred=True)
 
     # TODO: same as develop but for legacy reasons
     version('perfmodels', git=url)
@@ -101,7 +101,7 @@ class Coreneuron(CMakePackage):
             env['CC']  = 'tau_cc'
             env['CXX'] = 'tau_cxx'
         # for bg-q, our cmake needs mpi compilers as c, cxx compiler
-        elif 'bgq' in self.spec.architecture and spec.satisfies('+mpi'):
+        elif 'bgq' in spec.architecture and spec.satisfies('+mpi'):
             env['CC']  = spec['mpi'].mpicc
             env['CXX'] = spec['mpi'].mpicxx
 
@@ -109,6 +109,9 @@ class Coreneuron(CMakePackage):
                    '-DCMAKE_C_FLAGS=%s' % optflag,
                    '-DCMAKE_CXX_FLAGS=%s' % optflag,
                    '-DCMAKE_BUILD_TYPE=CUSTOM']
+
+        if 'bgq' in spec.architecture and '%xl' in spec:
+            options.append('-DCMAKE_BUILD_WITH_INSTALL_RPATH=1')
 
         if spec.satisfies('+tests'):
             options.extend(['-DUNIT_TESTS=ON',

--- a/packages/mod2c/package.py
+++ b/packages/mod2c/package.py
@@ -34,7 +34,9 @@ class Mod2c(CMakePackage):
     depends_on('cmake@2.8.12:', type='build')
 
     def cmake_args(self):
-        # todo: unnecessary method, check on cray though
-        options = ['-DCMAKE_C_COMPILER=%s' % spack_cc,
-                   '-DCMAKE_CXX_COMPILER=%s' % spack_cxx]
+        spec = self.spec
+        options = []
+        if 'bgq' in spec.architecture and '%xl' in spec:
+            options.append('-DCMAKE_BUILD_WITH_INSTALL_RPATH=1')
+
         return options

--- a/packages/reportinglib/package.py
+++ b/packages/reportinglib/package.py
@@ -57,7 +57,7 @@ class Reportinglib(CMakePackage):
             env['CC']  = 'tau_cc'
             env['CXX'] = 'tau_cxx'
         # for bg-q, our cmake needs mpi compilers as c, cxx compiler
-        elif 'bgq' in self.spec.architecture and spec.satisfies('+mpi'):
+        elif 'bgq' in self.spec.architecture:
             env['CC']  = spec['mpi'].mpicc
             env['CXX'] = spec['mpi'].mpicxx
 

